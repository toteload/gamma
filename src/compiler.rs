use crate::{
    ast::{NodeId, NodeIdGenerator},
    error::Error,
    ink_codegen::CodeGenerator,
    parser::Parser,
    semantics::{validate_semantics, SemanticContext},
    source_location::SourceSpan,
    string_interner::{StringInterner, Symbol},
    type_check::type_check,
    types::{Signedness, Type, TypeInterner, TypeToken},
};
use std::collections::HashMap;

pub use crate::ink_codegen::{Options};

pub struct Context {
    pub id_generator: NodeIdGenerator,

    pub symbols: StringInterner,
    pub type_tokens: TypeInterner,
    pub type_table: HashMap<Symbol, TypeToken>,

    pub spans: HashMap<NodeId, SourceSpan>,
    pub types: HashMap<NodeId, TypeToken>,
}

impl Context {
    pub fn new() -> Self {
        let mut symbols = StringInterner::new();
        let mut type_tokens = TypeInterner::new();

        let type_table = HashMap::from([
            (
                symbols.add("i8"),
                type_tokens.add(Type::Int {
                    signedness: Signedness::Signed,
                    width: 8,
                }),
            ),
            (
                symbols.add("i16"),
                type_tokens.add(Type::Int {
                    signedness: Signedness::Signed,
                    width: 16,
                }),
            ),
            (
                symbols.add("i32"),
                type_tokens.add(Type::Int {
                    signedness: Signedness::Signed,
                    width: 32,
                }),
            ),
            (
                symbols.add("i64"),
                type_tokens.add(Type::Int {
                    signedness: Signedness::Signed,
                    width: 64,
                }),
            ),
            (
                symbols.add("u8"),
                type_tokens.add(Type::Int {
                    signedness: Signedness::Unsigned,
                    width: 8,
                }),
            ),
            (
                symbols.add("u16"),
                type_tokens.add(Type::Int {
                    signedness: Signedness::Unsigned,
                    width: 16,
                }),
            ),
            (
                symbols.add("u32"),
                type_tokens.add(Type::Int {
                    signedness: Signedness::Unsigned,
                    width: 32,
                }),
            ),
            (
                symbols.add("u64"),
                type_tokens.add(Type::Int {
                    signedness: Signedness::Unsigned,
                    width: 64,
                }),
            ),
            (symbols.add("void"), type_tokens.add(Type::Void)),
            (symbols.add("bool"), type_tokens.add(Type::Bool)),
        ]);

        Context {
            id_generator: NodeIdGenerator::new(),
            symbols,
            type_tokens,
            type_table,
            spans: HashMap::new(),
            types: HashMap::new(),
        }
    }

    pub fn compile(&mut self, source: &str, options: &Options) -> Result<String, Vec<Error>> {
        let mut parser = Parser::new(
            source,
            &mut self.symbols,
            &mut self.id_generator,
            &mut self.spans,
        );

        let items = parser.parse_items().map_err(|e| vec![e])?;

        let semantic_context = SemanticContext {
            symbols: &self.symbols,
            spans: &self.spans,
        };

        validate_semantics(&semantic_context, &items)?;

        type_check(
            &items,
            &mut self.type_tokens,
            &mut self.types,
            &mut self.type_table,
        )?;

        let ctx = inkwell::context::Context::create();
        let mut codegen = CodeGenerator::new(&ctx, &self.symbols, &self.types, &self.type_tokens);
        let output = codegen.compile(&items, options).map_err(|e| vec![e])?;

        Ok(output)
    }
}
