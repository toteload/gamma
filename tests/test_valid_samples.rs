use gamma::compiler::{{Context, Options, OutputTarget}};
use insta::assert_snapshot;
use std::fs;

#[test]
fn basic_arithmetic() {
    let contents = fs::read_to_string("tests/valid_samples/basic_arithmetic.gamma").unwrap();

    let mut context = Context::new();
    let result = context.compile(&contents, &Options { optimize: false, output: OutputTarget::LlvmIr });

    let Ok(output) = result else {
        let Err(errors) = result else { unreachable!() };

        for e in &errors {
            e.print(&context);
        }

        panic!("Compilation of sample \"basic_arithmetic\" resulted in {} error(s)", errors.len()); 
    };

    assert_snapshot!("basic_arithmetic", output);
}

#[test]
fn basic_main() {
    let contents = fs::read_to_string("tests/valid_samples/basic_main.gamma").unwrap();

    let mut context = Context::new();
    let result = context.compile(&contents, &Options { optimize: false, output: OutputTarget::LlvmIr });

    let Ok(output) = result else {
        let Err(errors) = result else { unreachable!() };

        for e in &errors {
            e.print(&context);
        }

        panic!("Compilation of sample \"basic_main\" resulted in {} error(s)", errors.len()); 
    };

    assert_snapshot!("basic_main", output);
}

#[test]
fn nested_loop() {
    let contents = fs::read_to_string("tests/valid_samples/nested_loop.gamma").unwrap();

    let mut context = Context::new();
    let result = context.compile(&contents, &Options { optimize: false, output: OutputTarget::LlvmIr });

    let Ok(output) = result else {
        let Err(errors) = result else { unreachable!() };

        for e in &errors {
            e.print(&context);
        }

        panic!("Compilation of sample \"nested_loop\" resulted in {} error(s)", errors.len()); 
    };

    assert_snapshot!("nested_loop", output);
}

#[test]
fn sum_loop() {
    let contents = fs::read_to_string("tests/valid_samples/sum_loop.gamma").unwrap();

    let mut context = Context::new();
    let result = context.compile(&contents, &Options { optimize: false, output: OutputTarget::LlvmIr });

    let Ok(output) = result else {
        let Err(errors) = result else { unreachable!() };

        for e in &errors {
            e.print(&context);
        }

        panic!("Compilation of sample \"sum_loop\" resulted in {} error(s)", errors.len()); 
    };

    assert_snapshot!("sum_loop", output);
}
#[test]
fn basic_arithmetic_optimized_assembly() {
    let contents = fs::read_to_string("tests/valid_samples/basic_arithmetic.gamma").unwrap();

    let mut context = Context::new();
    let result = context.compile(&contents, &Options { optimize: true, output: OutputTarget::Assembly });

    let Ok(output) = result else {
        let Err(errors) = result else { unreachable!() };

        for e in &errors {
            e.print(&context);
        }

        panic!("Compilation of sample \"basic_arithmetic_optimized_assembly\" resulted in {} error(s)", errors.len()); 
    };

    assert_snapshot!("basic_arithmetic_optimized_assembly", output);
}

#[test]
fn basic_main_optimized_assembly() {
    let contents = fs::read_to_string("tests/valid_samples/basic_main.gamma").unwrap();

    let mut context = Context::new();
    let result = context.compile(&contents, &Options { optimize: true, output: OutputTarget::Assembly });

    let Ok(output) = result else {
        let Err(errors) = result else { unreachable!() };

        for e in &errors {
            e.print(&context);
        }

        panic!("Compilation of sample \"basic_main_optimized_assembly\" resulted in {} error(s)", errors.len()); 
    };

    assert_snapshot!("basic_main_optimized_assembly", output);
}

#[test]
fn nested_loop_optimized_assembly() {
    let contents = fs::read_to_string("tests/valid_samples/nested_loop.gamma").unwrap();

    let mut context = Context::new();
    let result = context.compile(&contents, &Options { optimize: true, output: OutputTarget::Assembly });

    let Ok(output) = result else {
        let Err(errors) = result else { unreachable!() };

        for e in &errors {
            e.print(&context);
        }

        panic!("Compilation of sample \"nested_loop_optimized_assembly\" resulted in {} error(s)", errors.len()); 
    };

    assert_snapshot!("nested_loop_optimized_assembly", output);
}

#[test]
fn sum_loop_optimized_assembly() {
    let contents = fs::read_to_string("tests/valid_samples/sum_loop.gamma").unwrap();

    let mut context = Context::new();
    let result = context.compile(&contents, &Options { optimize: true, output: OutputTarget::Assembly });

    let Ok(output) = result else {
        let Err(errors) = result else { unreachable!() };

        for e in &errors {
            e.print(&context);
        }

        panic!("Compilation of sample \"sum_loop_optimized_assembly\" resulted in {} error(s)", errors.len()); 
    };

    assert_snapshot!("sum_loop_optimized_assembly", output);
}

