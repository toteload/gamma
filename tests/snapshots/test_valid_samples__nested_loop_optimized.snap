---
source: tests/test_valid_samples.rs
expression: output
---
; ModuleID = 'main'
source_filename = "main"

; Function Attrs: nocallback nofree nosync nounwind willreturn
declare ptr @llvm.stacksave() #0

; Function Attrs: nocallback nofree nosync nounwind willreturn
declare void @llvm.stackrestore(ptr) #0

define i64 @main() {
entry:
  %0 = call ptr @llvm.stacksave()
  br label %loop

loop:                                             ; preds = %endloop1, %entry
  %.0 = phi i64 [ 0, %entry ], [ %11, %endloop1 ]
  %1 = call ptr @llvm.stacksave()
  %2 = icmp eq i64 %.0, 10
  br i1 %2, label %then, label %otherwise

then:                                             ; preds = %loop
  %3 = call ptr @llvm.stacksave()
  call void @llvm.stackrestore(ptr %1)
  br label %endloop

otherwise:                                        ; preds = %loop
  br label %endif

endif:                                            ; preds = %otherwise
  %4 = alloca i64, align 8
  store i64 0, ptr %4, align 4
  br label %loop2

loop2:                                            ; preds = %endif3, %endif
  %5 = call ptr @llvm.stacksave()
  %6 = load i64, ptr %4, align 4
  %7 = icmp eq i64 %6, 10
  br i1 %7, label %then5, label %otherwise4

then5:                                            ; preds = %loop2
  %8 = call ptr @llvm.stacksave()
  call void @llvm.stackrestore(ptr %5)
  br label %endloop1

otherwise4:                                       ; preds = %loop2
  br label %endif3

endif3:                                           ; preds = %otherwise4
  %9 = load i64, ptr %4, align 4
  %10 = add i64 %9, 1
  store i64 %10, ptr %4, align 4
  call void @llvm.stackrestore(ptr %5)
  br label %loop2

endloop1:                                         ; preds = %then5
  %11 = add i64 %.0, 1
  call void @llvm.stackrestore(ptr %1)
  br label %loop

endloop:                                          ; preds = %then
  call void @llvm.stackrestore(ptr %0)
  ret i64 0
}

attributes #0 = { nocallback nofree nosync nounwind willreturn }

